### DISCUSSION TITLE
GSoC 2024 Projects ideas

### DISCUSSION THREAD

**Discussion by Prerna-0202** [CircuitVerse/CircuitVerse#4391] [General]
This thread has been created for gathering project ideas for GSoC 2024. You can suggest your ideas in this thread. Let's brainstorm and come up with innovative ideas together! 🚀

**Comment by rose2221** [comment]
I’d like to propose the idea for AI Chatbot integration with Circuitverse. Here is the doc for the same:-
[proposal_for_AIchatbot](https://docs.google.com/document/d/1pz02LycAAtK7bm5E28I67lelPdkPdLGDbeKQzZv-H2w/edit?usp=sharing)

**Comment by Prajyot70** [comment]
Hi,
I'm Prajyot,
Here's my idea for Optimising Circuitverse Simulator For Mobile Devices.

Doc for the idea:- [Optimising Circuitverse Simulator For Mobile Devices](https://docs.google.com/document/d/1va4bFOxD_dhhUAFMZ0VTdoIwsVsZEVc3L6P0PwCJJ7U/edit?usp=drivesdk)

**Reply by tachyons** [reply]
Link is not accessible, could you share the summary here ?

**Reply by Prajyot70** [reply]
Apologies for the confusion, there was a problem with the accessibility of the doc, now it has been resolved. Still I'm posting the idea here.


_**Name:- Prajyot Bhusari**
**Github:- Prajyot70**_


# **_**Optimising Circuitverse Simulator For Mobile Devices**_**

In today's world, the usage of smart portable devices is enormous and is increasing day by day, the availability of your digital product on mobile devices is a necessity. According to sources on google about 60-65% of the google searches happen on mobile devices.
![Screenshot_20231220-013158_Chrome.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/bee6f263-a85a-4339-b4a9-860a22563fcc)


 Therefore, it's necessary to provide your user a good experience at first sight and make a good first impression, as the saying says " First impression is the last impression". A website which looks and performs badly on a mobile device can make a bad impression on the user and might make him leave the product forever. Also many people use smartphones as their primary device either because they can't afford a personal computer or because of its form factor. So to increase new user engagement and retention it is very important to make your website look and perform on mobile devices as or more good as it does on desktop computers.
That is the goal of this project, to optimise the Circuitverse Simulator for mobile devices to provide a better user experience by fixing existing bugs and improving the webview of the simulator on mobile devices.


**1. Introduce full screen mode for simulator**

Introducing full screen mode here means hiding the website address bar and navigation bar by covering it with the simulator. At present the navigation and web address bar acquire some space on the screen. By introducing full screen mode to the simulator on mobile devices, the small screen space can be used effectively. When the user launches the simulator, it will directly open in full screen mode to avoid hassle. The illustrations for the same are given below.

Existing:-
![Screenshot_20231219-231501_Chrome.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/39527fda-be04-4080-bc5d-1529e882d945)



Proposed:-
![Picsart_23-12-19_19-17-34-744.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/4a7e0b9f-0c74-4260-b53a-64c3ef89546a)



**2. Enable portrait mode for simulator on mobile/portable devices**

Enabling the portrait mode for circuitverse simulator on mobile devices will give an extra edge as this feature is only available on mobile devices and will be a plus point as people who are comfortable in using their smartphones in portrait mode, which most do, will engage better with the simulator on their mobile device.
The steps to achieve this are as follows.
Enabling the portrait mode for circuitverse simulator on mobile devices, which is currently disabled.
Changing element size and positioning to deliver a better user experience.

The illustrations for the following are given below.

Existing:-
![Screenshot_20231218-224650_Chrome.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/419ad97c-42f3-42ec-9d23-8a74afac1d56)



Proposed:-
![Picsart_23-12-19_17-48-18-557.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/a85c2fad-24f1-4ee9-a145-77fc64c9df02)



**3. Fix existing bugs**

There are some bugs in the existing circuitverse simulator, which hinders its usability on mobile devices.
Examples of bugs in the existing Circuitverse simulator are as follows, the illustrations for each bug are given just after the bug statement.

- Menu options going out of screen and can't be accessed by scrolling or any other way
![Picsart_23-12-20_02-09-00-417.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/e36fc53c-64c4-4559-80df-548740b6a779)

        

- Button going out of screen and isn't accessible by any means
![Picsart_23-12-20_02-04-32-812.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/91be17ca-6fca-4cf2-87e3-4805cfde6c7c)



- Elements at bottom of screen touching the edge of display and hard to select
![Picsart_23-12-20_02-12-48-820.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/858a4a2c-6631-413a-a72c-514ee1c9789f)



- zoom in/out and slide gestures which does not function properly on mobile devices

- Drop down menu covering other elements of the Circuitverse Simulator
![Picsart_23-12-20_02-16-24-442.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/08c42c55-019a-4c42-b55c-41a3c6281d83)



Following are the fixes for the above mentioned bugs

- Repositioning the menu layout to always stay in the middle of the screen, no matter which option is chosen from the elements. The triangular pointer will move up and down according to the element selected, not the whole menu layout.

In portrait mode:-
![Picsart_23-12-19_18-57-24-580.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/e449185a-4b34-4a83-b394-bf8c2615e161)


 
In Landscape mode:-
![Picsart_23-12-19_19-22-10-254.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/fe2687e9-b046-4285-bc5d-abf591c87580)




- The above solution can be implemented for the second bug as well as they are of the same type, just the area of effect is different, in first case it's a menu option and in second case a button.

- Repositioning the elements by adjusting the distance of an element from other elements and from  screen edges and align them symmetrically to improve their visibility and accessibility.

In portrait mode:-
![Picsart_23-12-19_17-48-18-557.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/07ba8883-fa3e-4286-9bf9-0465398c280d)



In Landscape mode:-
![Picsart_23-12-19_19-12-49-329.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/c343fbb8-d673-42dc-8a90-79bec7dcd55f)



- Improve zoom in/out and slide gestures on mobile devices

- Redesigning the menu layout is the solution to avoid overlapping of two elements in the Circuitverse Simulator

**4. Reposition and resize elements
 hiding lightly used/ less important icons and highlighting frequently used ones
**
Resizing elements help improve their visibility and accessibility, it is difficult to interact with a small sized element on a smaller sized screen.
Repositioning the control elements of the simulator to a mobile friendly interface, which not just makes them look good but also improves their usability. This can be achieved by the following ways and illustrations for the same are given below.
Choosing right size for the elements according to the screen size of the device
Positioning the elements such that they do not overlap or collide with other elements and also taking in to consideration their accessibility
Design layout for the elements in order to improve the appearance of the simulator and efficient usage of screen size
Using drawer menu to hide less frequently used elements to make screen available for the actual circuit.

In Landscape mode:-
![Picsart_23-12-19_19-17-34-744.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/f3548e2e-4887-4242-bda8-a9dd295df36f)



In portrait mode:-
![Picsart_23-12-19_17-52-44-342.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/1055b9ed-6059-4217-93eb-e6d5b32f0010)




**5. Redesign menu layout to improve user experience on mobile devices**

Redesigning the menu layout from scratch to improve its visibility, accessibility and overall user experience. Illustrations for the same are given below.

Existing:-
![Screenshot_20231220-005347_Chrome.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/0dabd5d2-77de-4f27-9610-65bbdeaef70b)



Proposed:-
In Landscape mode:-
![Picsart_23-12-20_02-01-18-908.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/8c267e2d-a91d-4b8b-96a7-ae93e2a80ed4)



In portrait mode:-
![Picsart_23-12-20_01-56-25-086.jpg](https://github.com/CircuitVerse/CircuitVerse/assets/140729090/d4d65242-72b5-4c89-bbce-e7916390bbe0)



By implementing the above mentioned solutions the engagement and user experience of the Circuitverse logic design simulator can be enhanced.

**Thank You!**

**Comment by niladrix719** [comment]
# Updating The Desktop Application

The primary objective of this project is to update the existing Electron desktop application for CircuitVerse or, alternatively, create a new repository and find a way to synchronise it with the `cv-frontend-vue` repository. The synchronisation might be achieved through the use of Git submodules.


## Updating / Creating The Repository

- We can use [electron-vite](https://electron-vite.org) configed with **VueJS + Typescript** for creating the repository.
- The new changes in the **main simulator** can be updated in the `cv-frontend-vue` repository to keep it **up to date** and ready for synchronisation with the desktop application.

## Synchronisation

- Using of **Git submodules** or finding alternatives for the Synchronisation.
- `cv-frontend-vue` repository will be synchronised with both the Circuitverse main repository and the Desktop repository.

Synchronisation flow chart:

```mermaid
graph LR
C(Circuitverse Main Repository)
B(Desktop Applicaton) --> D{cv-frontend-vue}
C --> D
```

## Benefits of Desktop Application

- Users will gain the ability to access CircuitVerse even in **offline mode**, eliminating the dependency on an internet connection.
- ElectronJS will be utilised to build Desktop application that seamlessly run on **MacOS, Linux, and Windows** operating systems.
- Will Simplify file import and export processes, providing users with easy access to manage their circuit designs effortlessly.

**Reply by Arnabdaz** [reply]
Electron is not the only option, we should also look into the possibility of using [Tauri](https://tauri.app/) instead of Electron if possible.

**Reply by niladrix719** [reply]
Yes, if our goal is to create a lightweight application with minimal memory consumption, Tauri appears to be a good option.

**Comment by adarsh7raj** [comment]
I'd like to propose a idea for Enhanced user dashboard and analytics. Here is the document for that:

https://docs.google.com/document/d/1ydugu7rWhXJJMSNduQc_V2QPLejiUkxQXSFMxGpdrnM/edit?usp=sharing

**Comment by Mayank77maruti** [comment]
<h1>Enhancing Desktop Application Development with Tauri</h1>
<p>The primary goal of this project is to establish a new repository and devise a method to synchronize it with the <mark>cv-frontend-vue</mark> repository while leveraging Tauri for building the application. The synchronization process could potentially be achieved through the utilization of Git submodules.</p>

<h1>Repository Creation/Update</h1>
<ul>
  <li>We can utilize <a href="https://tauri.app/">Tauri</a> along with VueJS and Typescript to create the repository.</li>
  <li>It's essential to integrate recent updates from the main simulator into the cv-frontend-vue repository to ensure it remains current and ready for synchronization with the desktop application.</li>
</ul>

<h1>Synchronization</h1>
<ul>
  <li>We can employ Git submodules or explore alternative methods for synchronization.</li>
  <li>The cv-frontend-vue repository will be synchronized with both the Circuitverse main repository and the Desktop repository.</li>
</ul>

<h1>Advantages of Desktop Application Development with Tauri</h1>
<ul>
  <li>Native Performance: Tauri provides access to native APIs, resulting in high-performance applications.</li>
  <li>Compact Bundle Size: Tauri applications have a small footprint, ensuring efficiency and speed.</li>
  <li>Cross-Platform Compatibility: Tauri supports building applications for multiple operating systems, including Windows, macOS, and Linux.</li>
<li>Offline Mode Support: Users will gain the ability to access CircuitVerse even in offline mode, eliminating the dependency on an internet connection.</li>
</ul>

**Comment by JamesCartar** [comment]
I would like to propose an enhancement to the user interface. Currently, most of the editor components on the CircuitVerse page are draggable and can be minimized, maximized, and placed according to the user’s preference. While this offers flexibility, it can also lead to a cluttered workspace.

I suggest that we categorize and combine related components and place them in sidebars on the left and right. This would streamline the interface and improve the user experience by providing a more organized workspace.

Additionally, when asking for project names or details, instead of using the browser’s default alert, we could implement a custom pop-up box. This would not only make the interface more consistent but also give us the opportunity to enhance user interaction with these prompts.

I believe these changes would significantly improve the usability and aesthetics of CircuitVerse, making it even more appealing and user-friendly.

**Comment by Igneel0601** [comment]
I would like to suggest the following changes:

--> Making UI consistent throughout the website.
--> Include simple and easy to understand examples like half adder, multiplexer, ring counter, etc.
--> I personally think that homepage is a little large. Making it concise.
--> Name for components in simulator instead of seeing name only when you hover. Since these components might look a little different.
--> blog.circuitverse main icon doesn't redirect you to homepage of circuit verse. Either making it go back to home or changing it a little. 
       Same with learn.circuitverse.
--> Sending recent circuits and its pagination to a different page
--> Improving side bar of learn.circuitverse
--> Sign up process is faulty. It takes any made up email address without verifying, we can add a verification system.
--> In docs.circuitverse instead of sidebar, add the icon  to top left and include the logo.
--> Scaling issue in docs.circuit verse. Go to https://docs.circuitverse.org/#/chapter2/1gettingstarted and scroll down till Login to CircuitVerse 
      User Account is on top and then open and close the sidebar.
--> In mobile app, featured circuits are same as editor picks.
--> In mobile app, you cannot click on buttons of email us, slack and github.
--> Also including a report to us section in app, since there is a chance that users find a bug that the developers might miss

**Comment by senbo1** [comment]
Hi, I am Harsh, here is my solution for CircuitVerse Real Time Collaboration 
https://docs.google.com/document/d/1debQSIvUHp9dp4G5mVBKHhPaKCLo2Hs6YawGAhHeF_c/edit?usp=sharing

**Comment by s2ahil** [comment]
using next js in circuit verse  for frontend and backend.

**Comment by DhairyaMajmudar** [comment]
# Project Idea For Vue Simulator Improvement 


Attaching a Google Doc link for the idea

https://docs.google.com/document/d/1k_qstAWD7vDVDwiKbbkH_dNwfy2KRe03MdM6fBsdrOE/edit?usp=sharing

**Comment by EmAreAitch** [comment]
# Project Idea for a Educational Board Game of Digital Circuits

Here is the google docs link for my idea proposal - [Digital Circuit Game](https://docs.google.com/document/d/1FKIpxpFALG76Vf9m1OAD4eT8dK0UmWfZSdld2iKvScU/edit?usp=sharing)

**Comment by tachyons** [comment]
I just wante do clarify that this thread is not to submit project proposals. That will be done during gsoc period. Here we are looking for ideas in concise format which can be considered to be included in the ideas list.

**Comment by EmAreAitch** [comment]
Yeah thanks for clarifying. Mine is also an idea only. I also got confused looking at other people responses. Please don't mind me.

> # Proposal for a Educational Board Game for Digital Circuit
> Here is the google docs link for my proposal - [Digital Circuit Game](https://docs.google.com/document/d/1FKIpxpFALG76Vf9m1OAD4eT8dK0UmWfZSdld2iKvScU/edit?usp=sharing)

**Comment by MostlyKIGuess** [comment]
# Adding a new element "Transformers"
- To analyze certain analog electronic circuits transformers are a very basic aspect of understanding.
- Overall it's a very simple and an urgent idea in my opinion. I would love to work on this and I would need help :)
## Working of Transformers:
- https://www.fierceelectronics.com/electronics/what-a-transformer
- Basics first, the first prototype we can just use two voltage sources and the two number of coils for the both transformer to take into account about what will be the output voltage, we can start with a system block.
- Back end would just start with this that we would work with input and output voltage as a system block along with an amplifier.
## Design:
- We can keep two options in the starting that is for input and the number of coils so that the user can at least do the most basic part of transformer.

## Further Improvements:
- We can after the prototype is done move to center tapped transformer and analog rectifier circuits.
- So we can give examples of full wave and half wave rectifiers.

**Comment by APS-1926** [comment]
### Adding a new option for the users

We can add an AI recognition so that, a hand drawn circuit can be detected via an image input and then can be simulated according to user's choice, this can be both useful for educators and student to check their design in a much faster way as any time we go through some Idea , it is first drawn in physical note, this can be a time saving option too.

**Comment by jayesh9747** [comment]
@tachyons  @Prerna-0202  Is this discussion open for 2025?

**Comment by Meenatchi2005** [comment]
Hi, I'm Meenatchi, a student from India currently learning web development. I have basic knowledge of HTML, CSS, JavaScript, and Python. I'm interested in the AI Chatbot project for GSoC 2025 and would love to start contributing. Could you please guide me on how to begin?
